format-xls/src/main/java/io/cdap/plugin/format/xls/input/XlsRowConverter.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright © 2024 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.format.xls.input;
+
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import org.apache.poi.ss.usermodel.Cell;
+import org.apache.poi.ss.usermodel.CellType;
+import org.apache.poi.ss.usermodel.DataFormatter;
+import org.apache.poi.ss.usermodel.DateUtil;
+import org.apache.poi.ss.usermodel.FormulaEvaluator;
+import org.apache.poi.ss.usermodel.Row;
+
+import java.util.List;
+
+/**
+ * Converts a row of XLS cells to a StructuredRecord.
+ */
+public class XlsRowConverter {
+  private final FormulaEvaluator evaluator;
+  private static final DataFormatter dataFormatter = new DataFormatter();
+
+  XlsRowConverter(FormulaEvaluator evaluator) {
+    this.evaluator = evaluator;
+  }
+
+  /**
+   * Converts a row of XLS cells to a StructuredRecord.
+   * Returns null if the row is null or empty.
+   */
+  public StructuredRecord.Builder convert(Row row, Schema outputSchema) {
+    if (row == null) {
+      return null;
+    }
+    boolean isRowEmpty = true;
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    List<Schema.Field> fields = outputSchema.getFields();
+    for (int cellIndex = 0; cellIndex < row.getLastCellNum(); cellIndex++) {
+      if (cellIndex >= fields.size()) {
+        throw new IllegalArgumentException(
+                String.format("Schema contains fewer fields than the number of columns in the excel file. " +
+                        "Schema fields: %s, Excel columns: %s", fields.size(), row.getLastCellNum()));
+      }
+      Cell cell = row.getCell(cellIndex, Row.MissingCellPolicy.RETURN_BLANK_AS_NULL);
+      if (cell == null) {
+        // Blank cells are skipped, builder will set null for the field, no processing needed.
+        continue;
+      }
+      Schema.Field field = fields.get(cellIndex);
+      Schema.Type type = field.getSchema().isNullable() ?
+              field.getSchema().getNonNullable().getType() : field.getSchema().getType();
+      Object cellValue;
+      switch (type) {
+        case STRING:
+          cellValue = getCellAsString(cell);
+          break;
+        case DOUBLE:
+          cellValue = getCellAsDouble(cell);
+          break;
+        case BOOLEAN:
+          cellValue = getCellAsBoolean(cell);
+          break;
+        default:
+          throw new IllegalArgumentException(
```

Comment:
this should be checked during plugin validation if it is not already happening. For cases like this when we don't expect the situation to ever happen, we usually add a comment that it's not expected and throw an IllegalStateException

Reply:
Yup, this is not suppose to happen.
I have added comments for that.

Reply:
nit: throw IllegalStateException instead of IllegalArgumentException, since hitting this is a problem in the code and not a problem with the input argument